wagtailio/newsletter/migrations/0005_wagtail_newsletter_data.py

```diff
@@ -0,0 +1,42 @@
+# Generated by Django 5.1.6 on 2025-04-16 13:02
+
+from django.db import migrations
+
+
+def convert_body_to_streamfield(apps, schema_editor):
+    from bs4 import BeautifulSoup
+
+    NewsletterPage = apps.get_model("newsletter", "NewsletterPage")
+    for page in NewsletterPage.objects.all():
+        soup = BeautifulSoup(page.intro, "html.parser")
+        preview = soup.get_text(strip=True)
+        if preview:
+            page.preview = preview
+            page.save()
+
+        if page.body:
+            page.content = [{"type": "rich_text", "value": page.body}]
+            page.save()
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("newsletter", "0004_wagtail_newsletter_model"),
+    ]
+
+    operations = [
+        migrations.RunPython(convert_body_to_streamfield, migrations.RunPython.noop),
+        migrations.RemoveField(
+            model_name="newsletterpage",
+            name="body",
+        ),
+        migrations.RemoveField(
+            model_name="newsletterpage",
+            name="intro",
+        ),
+        migrations.RenameField(
+            model_name="newsletterpage",
+            old_name="content",
+            new_name="body",
+        ),
```

Comment:
I'd recommend splitting the `RemoveField` operations to another migration. The problem tends to happen if you have both a data migration operation (in this case `RunPython(convert_body_to_streamfield)`) and a schema migration operation in a single migration. Migrations are run in a single transaction (if the backend supports DDL transactions), and the database may not allow data manipulation in a DDL transaction.

Reply:
Fair point. I've also noticed that images were not getting preserved in the data migration so I've fixed that too.